fuse_core/include/fuse_core/util.hpp

```diff
@@ -150,6 +151,178 @@ Eigen::Matrix<T, 2, 2, Eigen::RowMajor> rotationMatrix2D(const T angle)
+        (2.0 * qx * qx + 2.0 * qy * qy - 1.0));
+      jacobian_map(0, 1) =
+        -((2.0 * qw) / (2.0 * qx * qx + 2.0 * qy * qy - 1.0) -
+        (4.0 * qx * (2.0 * qw * qx + 2.0 * qy * qz)) / std::pow((2.0 * qx * qx + 2.0 * qy * qy - 1.0), 2.0)) /
+        (std::pow((2.0 * qw * qx + 2.0 * qy * qz), 2.0) / std::pow((2.0 * qx * qx + 2.0 * qy * qy - 1.0), 2.0) + 1.0);
+      jacobian_map(0, 2) =
+        -((2.0 * qz) / (2.0 * qx * qx + 2.0 * qy * qy - 1.0) -
+        (4.0 * qy * (2.0 * qw * qx + 2.0 * qy * qz)) / std::pow((2.0 * qx * qx + 2.0 * qy * qy - 1.0), 2.0)) /
+        (std::pow((2.0 * qw * qx + 2.0 * qy * qz), 2.0) / std::pow((2.0 * qx * qx + 2.0 * qy * qy - 1.0), 2.0) + 1.0);
+      jacobian_map(0, 3) =
+        -(2.0 * qy) /
+        ((std::pow((2.0 * qw * qx + 2.0 * qy * qz), 2.0) / std::pow((2.0 * qx * qx + 2.0 * qy * qy - 1.0), 2.0) +
+        1.0) *
+        (2.0 * qx * qx + 2.0 * qy * qy - 1.0));
+
+      jacobian_map(1, 0) = (2.0 * qy) / std::sqrt(1.0 - std::pow((2.0 * qw * qy - 2.0 * qx * qz), 2.0));
+      jacobian_map(1, 1) = -(2.0 * qz) / std::sqrt(1.0 - std::pow((2.0 * qw * qy - 2.0 * qx * qz), 2.0));
+      jacobian_map(1, 2) = (2.0 * qw) / std::sqrt(1.0 - std::pow((2.0 * qw * qy - 2.0 * qx * qz), 2.0));
+      jacobian_map(1, 3) = -(2.0 * qx) / std::sqrt(1.0 - std::pow((2.0 * qw * qy - 2.0 * qx * qz), 2.0));
+
+      jacobian_map(2, 0) =
+        -(2.0 * qz) /
+        ((std::pow((2.0 * qw * qz + 2.0 * qx * qy), 2.0) / std::pow((2.0 * qy * qy + 2.0 * qz * qz - 1.0), 2.0) +
+        1.0) *
+        (2.0 * qy * qy + 2.0 * qz * qz - 1.0));
+      jacobian_map(2, 1) =
+        -(2.0 * qy) /
+        ((std::pow((2.0 * qw * qz + 2.0 * qx * qy), 2.0) / std::pow((2.0 * qy * qy + 2.0 * qz * qz - 1.0), 2.0) +
+        1.0) *
+        (2.0 * qy * qy + 2.0 * qz * qz - 1.0));
+      jacobian_map(2, 2) =
+        -((2.0 * qx) / (2.0 * qy * qy + 2.0 * qz * qz - 1.0) -
+        (4.0 * qy * (2.0 * qw * qz + 2.0 * qx * qy)) / std::pow((2.0 * qy * qy + 2.0 * qz * qz - 1.0), 2.0)) /
+        (std::pow((2.0 * qw * qz + 2.0 * qx * qy), 2.0) / std::pow((2.0 * qy * qy + 2.0 * qz * qz - 1.0), 2.0) + 1.0);
+      jacobian_map(2, 3) =
+        -((2.0 * qw) / (2.0 * qy * qy + 2.0 * qz * qz - 1.0) -
+        (4.0 * qz * (2.0 * qw * qz + 2.0 * qx * qy)) / std::pow((2.0 * qy * qy + 2.0 * qz * qz - 1.0), 2.0)) /
+        (std::pow((2.0 * qw * qz + 2.0 * qx * qy), 2.0) / std::pow((2.0 * qy * qy + 2.0 * qz * qz - 1.0), 2.0) + 1.0);
+    }
+  }
+}
+
+/**
+ * @brief Compute product of two quaternions and the function jacobian
+ * TODO(giafranchini): parametric jacobian computation? Atm this function is only used in 
+ * normal_prior_pose_3d cost function. There we only need the derivatives wrt quaternion W,
+ * so at the time we are only computing the jacobian wrt W
+ * 
+ * @param[in] z Pointer to the first quaternion array (4x1)
+ * @param[in] w Pointer to the second quaternion array (4x1)
+ * @param[in] z Pointer to the first quaternion array (4x1) 
```

Comment:
```suggestion
 * @param[in] zw Pointer to the output quaternion array (4x1) that will be populated with the result of z * w
```
Or whatever description you want